tests/PrefixCacheBootstrapperTest.php

```diff
@@ -0,0 +1,382 @@
+<?php
+
+declare(strict_types=1);
+
+use Illuminate\Cache\CacheManager;
+use Illuminate\Support\Facades\Event;
+use Stancl\Tenancy\Events\TenancyEnded;
+use Stancl\Tenancy\Tests\Etc\CacheService;
+use Stancl\Tenancy\Events\TenancyInitialized;
+use Stancl\Tenancy\Listeners\BootstrapTenancy;
+use Stancl\Tenancy\Listeners\RevertToCentralContext;
+use Stancl\Tenancy\Tests\Etc\SpecificCacheStoreService;
+use Stancl\Tenancy\Bootstrappers\PrefixCacheTenancyBootstrapper;
+
+beforeEach(function () {
+    config([
+        'tenancy.bootstrappers' => [
+            PrefixCacheTenancyBootstrapper::class
+        ],
+        'cache.default' => $cacheDriver = 'redis',
+        'cache.stores.redis2' => config('cache.stores.redis'),
+    ]);
+
+    PrefixCacheTenancyBootstrapper::$tenantCacheStores = [$cacheDriver];
+    PrefixCacheTenancyBootstrapper::$prefixGenerator = null;
+
+    config('tenancy.cache.manager')::$addTags = false;
+
+    Event::listen(TenancyInitialized::class, BootstrapTenancy::class);
+    Event::listen(TenancyEnded::class, RevertToCentralContext::class);
+});
+
+afterEach(function () {
+    PrefixCacheTenancyBootstrapper::$tenantCacheStores = [];
+    PrefixCacheTenancyBootstrapper::$prefixGenerator = null;
+});
+
+test('Tenancy overrides CacheManager', function() {
+    $tenancyCacheManager = config('tenancy.cache.manager');
+
+    expect(app('cache')::class)->toBe($tenancyCacheManager);
+    expect(app(CacheManager::class)::class)->toBe($tenancyCacheManager);
+
+    tenancy()->initialize(Tenant::create(['id' => 'first']));
+
+    expect(app('cache')::class)->toBe($tenancyCacheManager);
+    expect(app(CacheManager::class)::class)->toBe($tenancyCacheManager);
+
+    tenancy()->initialize(Tenant::create(['id' => 'second']));
+
+    expect(app('cache')::class)->toBe($tenancyCacheManager);
+    expect(app(CacheManager::class)::class)->toBe($tenancyCacheManager);
+
+    tenancy()->end();
+
+    expect(app('cache')::class)->toBe($tenancyCacheManager);
+    expect(app(CacheManager::class)::class)->toBe($tenancyCacheManager);
+});
+
+test('correct cache prefix is used in all contexts', function () {
+    $originalPrefix = config('cache.prefix');
+    $prefixBase = config('tenancy.cache.prefix_base');
+    $getDefaultPrefixForTenant = fn (Tenant $tenant) => $originalPrefix . $prefixBase . $tenant->getTenantKey();
+    $generatePrefixForTenant = function (Tenant $tenant) {
+        return app(PrefixCacheTenancyBootstrapper::class)->generatePrefix($tenant);
+    };
+
+    $expectCachePrefixToBe = function (string $prefix) {
+        expect($prefix . ':') // RedisStore suffixes prefix with ':'
+            ->toBe(app('cache')->getPrefix())
+            ->toBe(app('cache.store')->getPrefix());
```

Comment:
Should we use more stores here like we're using in `cache store prefix generation can be customized`?

Reply:
Makes sense. Also, there was another test that could be using more stores, so I added that. Now, it also makes sense to have both `redis` and `redis2` in `$tenantCacheStores` by default because we mostly want to check if prefixing works for multiple stores, and when we want to check if it works just for specific stores (e.g. in `only the stores specified in tenantCacheStores get prefixed`), we set `$tenantCacheStores` in that test – we end up with less `PrefixCacheTenancyBootstrapper::$tenantCacheStores = [...` calls, so the tests are a bit cleaner in the end (https://github.com/archtechx/tenancy/pull/1014/commits/52d10d36f96c31be590eb4b9d40163e4ab189cd2)

Reply:
Set the two stores in `beforeEach()`, and `[]` in `afterEach()`